utility_code/lib_py3/plugin_data.py

```diff
@@ -0,0 +1,295 @@
+    launch any new processes, which is easier to debug.
+
+    initializer can be set to a function that initializes any variables once
+    for each process worker, which for large static arguments is much faster
+    than putting them in additional_args which would copy them for each
+    iteration. initializer will be called with the arguments supplied in
+    init_args.
+    """
+    if not os.path.isdir(path):
+        raise ValueError("Expected path to be a folder")
+
+    parallel_args = []
+    for plugin_data_path in os.listdir(path):
+        if plugin_data_path.endswith(".json"):
+            full_path = os.path.join(path, plugin_data_path)
+            parallel_args.append((_create_plugindata_lambda, (full_path,), _finalize_plugindata_lambda, (autosave,), func, err_func, additional_args))
+
+    yield from process_in_parallel(parallel_args, num_processes=num_processes, initializer=initializer, initargs=initargs)
+
+class PluginData(NbtPathDebug):
+    """Plugin data loaded alongside player data."""
+
+    def __init__(self, json_data=None, json_path=None):
+        """Load a plugin data from json.
+
+        Must be saved from wherever the json was loaded from for changes to apply.
+        """
+        self._path = json_path
+        self._data = json_data
+
+        if self._path is not None and self._data is None:
+            with open(self._path, 'r') as fp:
+                self._data = json.load(fp)
+                fp.close()
+
+        if self._data is None:
+            raise ValueError("No json data or path was provided")
+
+        self.nbt_path_init(None, None, self, None)
+
+    def get_debug_str(self):
+        return str(self)
+
+    def save(self, json_path=None):
+        """Save json data.
+
+        Always returns the json data.
+        If loaded from a path, or a path is provided, saves to that path.
+        """
+        if json_path is None:
+            json_path = self._path
+
+        if json_path:
+            with open(json_path, 'w') as fp:
+                json.dump(
+                    self._data,
+                    fp,
+                    ensure_ascii=False,
+                    indent=2,
+                    separators=(',', ': '),
+                    sort_keys=False
+                )
+                fp.write("\n")
+
+        return self._data
+
+    def graves(self):
+        """Get the graves stored on a player, if they exist."""
+        return MonumentaGraves(self._data.get("MonumentaGravesV2", {}), self)
+
+    def __repr__(self):
+        return f'PluginData({os.path.basename(self._path)!r})'
+
+class MonumentaGraves(NbtPathDebug):
+    """A collection of items preserved from death or carelessness."""
+
+    def __init__(self, json_data, parent):
+        self._data = json_data
+
+        self.nbt_path_init(None, parent, parent.root if parent is not None and parent.root is not None else self, None)
+
+    def get_debug_str(self):
+        return str(self)
```

Comment:
There's no `__str__` method defined; I'm not sure what this will display.

Reply:
None of the other ones do either, but apparently __repr__() seems to be what gets returned here? Should I use __str__ instead? I don't really understand the difference

Reply:
If it works, I'll take it :)